### PR TITLE
#640 Tests for SolidMap ctors

### PR DESCRIPTION
**Pull request by flslkxtc**
Issue #640. 
This pull request covers following SolidMap constructors with tests:
- `public SolidMap(final Map<X, Y> map, final Map.Entry<X, Y>... list)`
- `public <Z> SolidMap(final Func<Z, X> key, final Func<Z, Y> value, final Map<X, Y> map, final Iterable<Z> list)`
- `public <Z> SolidMap(final Func<Z, X> key, final Func<Z, Y> value, final Iterable<Z> list)`

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#669](https://github.com/yegor256/cactoos/issues/669) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#669](https://github.com/yegor256/cactoos/issues/669) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C63314D6Z/481129756). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

File: src/test/java/org/cactoos/map/SolidMapTest.java [hunk 1]
```diff
@@ -81,4 +86,71 @@ public void mapsToSameObjects() throws Exception {
             map.get(0), Matchers.equalTo(map.get(0))
         );
     }
+
+    @Test
+    @SuppressWarnings({"unchecked", "MismatchedQueryAndUpdateOfCollection"})
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/map/SolidMapTest.java]: @flslkxtc Is this suppress needed here? Why?

File: src/test/java/org/cactoos/map/SolidMapTest.java [hunk 2]
```diff
@@ -81,4 +86,71 @@ public void mapsToSameObjects() throws Exception {
...
+        final SolidMap<Integer, Integer> toadd = new SolidMap<>(
+            new MapEntry<>(2, 12),
+            new MapEntry<>(3, 13)
+        );
+        final Map.Entry<Integer, Integer>[] array =
+            (Map.Entry<Integer, Integer>[]) new Map.Entry<?, ?>[toadd.size()];
+        toadd.entrySet().toArray(array);
+        MatcherAssert.assertThat(
+            "Does not extends existing map with array of entries",
+            new SolidMap<Integer, Integer>(map, array).size(),
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/map/SolidMapTest.java]: @flslkxtc This ctor accepts a ``vararg`` array (``...``) -- isn't it possible to instantiate ``Map.Entry`` on the spot and remove ``toadd`` and ``array``? 

At least, if this is not possible, get rid of ``toadd`` and just say 

```java
final Map.Entry<Integer, Integer>[] array =
    (Map.Entry<Integer, Integer>[]) new Map.Entry<?, ?>[toadd.size()];
new SolidMap<>(
    new MapEntry<>(2, 12),
   new MapEntry<>(3, 13)
).entrySet().toArray(
    array
);
```

File: src/test/java/org/cactoos/map/SolidMapTest.java [hunk 3]
```diff
@@ -81,4 +86,71 @@ public void mapsToSameObjects() throws Exception {
...
+    @Test
+    public void mapsIterableWithKeyFuncAndValueFunc() {
+        MatcherAssert.assertThat(
+            "Functions are not applied to key and value",
+            new SolidMap<>(
+                key -> new SubText(new TextOf(key), 0, 1).asString(),
+                value -> new UpperText(new TextOf(value)).asString(),
+                new IterableOf<>("aa", "bb")
+            ),
+            Matchers.allOf(
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/map/SolidMapTest.java]: @flslkxtc Can you also add a matcher for the size? To check that there are only 2 entries?

File: src/test/java/org/cactoos/map/SolidMapTest.java [hunk 4]
```diff
@@ -81,4 +86,71 @@ public void mapsToSameObjects() throws Exception {
...
+            "Function are not applied to entry",
+            new SolidMap<>(
+                entry -> new MapEntry<>(
+                    new SubText(new TextOf(entry), 0, 1).asString(),
+                    new UpperText(new TextOf(entry)).asString()
+                ),
+                new IterableOf<>("aa", "bb")
+            ),
+            Matchers.allOf(
+                Matchers.hasEntry(
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/map/SolidMapTest.java]: @flslkxtc Same here, a matcher for size?

**Comment by amihaiemil** [comment]
@flslkxtc ping

**Comment by flslkxtc** [comment]
@amihaiemil i've added tests for empty Iterables and fixed your remarks, please take a look again.

**Comment by amihaiemil** [comment]
@rultor good to merge

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by amihaiemil** [comment]
@yegor256 ping, can we merge this?

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job, as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by amihaiemil** [comment]
@ypshenychka ping

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
> @0crat quality good ([here](https://github.com/yegor256/cactoos/issues/669#issuecomment-367925610))

@ypshenychka Job `gh:yegor256/cactoos#669` is not assigned, can't get start time

**Comment by 0crat** [comment]
`QA` review completed: +15 points just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka), total is [+600](https://www.0crat.com/u/ypshenychka)

**Comment by 0crat** [comment]
`QA` review completed: +15 points just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka), total is [+600](https://www.0crat.com/u/ypshenychka)

**Comment by 0crat** [comment]
Order was finished, quality was "good": +20 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+925](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
> @0crat quality good ([here](https://github.com/yegor256/cactoos/issues/669#issuecomment-367925610))

@ypshenychka You can't do that, unless you have one of these roles: `ARC`, `PO`. Your current roles are: `QA`.

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +15 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+10779](https://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
This pull request [#669](https://github.com/yegor256/cactoos/issues/669) is assigned to @neonailol[/z](https://www.0crat.com/u/neonailol), here is [why](https://www.0crat.com/footprint/C63314D6Z/591804303). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by neonailol** [comment]
@yegor256 this job was assigned to me by mistake

**Comment by yegor256** [comment]
@0crat out

**Comment by 0crat** [comment]
> @0crat out ([here](https://github.com/yegor256/cactoos/issues/669#issuecomment-370436696))

@yegor256 @neonailol[/z](https://www.0crat.com/u/neonailol) resigned from [#669](https://github.com/yegor256/cactoos/issues/669), since the job is not in scope anymore

**Comment by 0crat** [comment]
> @0crat out ([here](https://github.com/yegor256/cactoos/issues/669#issuecomment-370436696))

@yegor256 The job [#669](https://github.com/yegor256/cactoos/issues/669) is now out of scope

**Pull request merged**